File: backend/app/repositories/assets.py
import uuid

from sqlalchemy import func, or_, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.models import Asset


class AssetRepository:
    def __init__(self, session: AsyncSession) -> None:
        self.session = session

    async def get(self, asset_id: uuid.UUID) -> Asset | None:
        return await self.session.get(Asset, asset_id)

    async def list_by_org(self, org_id: uuid.UUID, *, archived: bool = False) -> list[Asset]:
        """Active assets by default; pass archived=True for the archive view."""
        cond = Asset.archived_at.is_not(None) if archived else Asset.archived_at.is_(None)
        result = await self.session.execute(
            select(Asset)
            .where(Asset.org_id == org_id, cond)
            .order_by(Asset.created_at.desc())
        )
        return list(result.scalars().all())

    async def list_page(
        self,
        org_id: uuid.UUID,
        *,
        archived: bool = False,
        q: str | None = None,
        status: str | None = None,
        location: str | None = None,
        device_id: uuid.UUID | None = None,
        device_ids: list[uuid.UUID] | None = None,
        offset: int = 0,
        limit: int = 50,
    ) -> tuple[list[Asset], int]:
        """One page of the register. `device_id` is how a caller that wants one device's
        asset asks for it — pulling the whole register and finding it client-side turns into
        "search the first page only" the moment this is paged, and the asset panel would go
        blank for anyone past it."""
        from app.schemas.pagination import paginate

        cond = Asset.archived_at.is_not(None) if archived else Asset.archived_at.is_(None)
        stmt = select(Asset).where(Asset.org_id == org_id, cond)
        if device_id is not None:
            stmt = stmt.where(Asset.device_id == device_id)
        if status:
            stmt = stmt.where(Asset.status == status)
        if location is not None:
            # "" means the unassigned bucket, which the UI shows as its own option. Without
            # this case that filter would silently return nothing.
            stmt = stmt.where(
                Asset.location.is_(None) if location == "" else Asset.location == location
            )
        if device_ids:
            # For the devices table, which labels each row with its asset's state and
            # location. It used to fetch the whole register to label one page — 2,000 rows
            # read to annotate 50, and once the register is paged it would simply annotate
            # the wrong ones.
            stmt = stmt.where(Asset.device_id.in_(device_ids))
        if q:
            like = f"%{q.lower()}%"
            stmt = stmt.where(
                or_(
                    func.lower(Asset.name).like(like),
                    func.lower(Asset.asset_tag).like(like),
                    func.lower(Asset.serial_number).like(like),
                    func.lower(Asset.manufacturer).like(like),
                    func.lower(Asset.model).like(like),
                    func.lower(Asset.location).like(like),
                )
            )
        stmt = stmt.order_by(Asset.created_at.desc())
        rows, total, _, _ = await paginate(
            self.session, stmt, page=offset // max(1, limit) + 1, page_size=limit
        )
        return rows, total

    async def distinct_locations(self, org_id: uuid.UUID) -> list[str]:
        """Every location value in use, for the filter dropdown.

        Its own query because the dropdown must offer every location in the org, not the
        handful that happen to appear on the page you are looking at.
        """
        rows = (await self.session.execute(
            select(Asset.location)
            .where(Asset.org_id == org_id, Asset.archived_at.is_(None),
                   Asset.location.is_not(None), Asset.location != "")
            .distinct()
            .order_by(Asset.location)
        )).all()
        return [r[0] for r in rows]

    async def add(self, asset: Asset) -> Asset:
        self.session.add(asset)
        await self.session.flush()
        return asset

    async def delete(self, asset: Asset) -> None:
        await self.session.delete(asset)

File: portal/src/lib/api/assets.ts
import { apiClient } from "./client";
import type {
  Asset, AssetInput, AssetPassport, AssetStatus, AssetSummary, Page, PageParams,
} from "./types";

// Device passport — full lifecycle history + analytics for one asset.
export const getAssetPassport = (id: string) =>
  apiClient.get<AssetPassport>(`/assets/${id}/passport`).then((r) => r.data);

export const listAssets = (
  params: PageParams & {
    archived?: boolean; q?: string; status?: AssetStatus; location?: string;
  } = {},
) =>
  apiClient.get<Page<Asset>>("/assets", { params }).then((r) => r.data);

/** The asset record for one device, resolved server-side.
 *
 *  Callers used to fetch the whole register and find the row in the browser. Now that the
 *  register is paged that search would only ever look at the first page, so a device whose
 *  asset sat on page 2 would show as having no asset at all. */
/** Asset records for the devices on screen, keyed by device id.
 *
 *  The devices table labels each row with its asset's state and location. It used to fetch
 *  the whole register and join in the browser: 2,000 rows read to annotate 50, and once the
 *  register is paged it would annotate the wrong ones. */
export const assetsForDevices = async (deviceIds: string[]): Promise<Map<string, Asset>> => {
  const map = new Map<string, Asset>();
  if (deviceIds.length === 0) return map;
  const page = await apiClient
    .get<Page<Asset>>("/assets", {
      params: { device_ids: deviceIds, page_size: deviceIds.length },
      paramsSerializer: { indexes: null },
    })
    .then((r) => r.data);
  for (const a of page.items) if (a.device_id) map.set(a.device_id, a);
  return map;
};

export const getAssetForDevice = async (deviceId: string): Promise<Asset | null> => {
  const page = await apiClient
    .get<Page<Asset>>("/assets", { params: { device_id: deviceId, page_size: 1 } })
    .then((r) => r.data);
  return page.items[0] ?? null;
};

export const archiveAsset = (id: string) =>
  apiClient.post<Asset>(`/assets/${id}/archive`).then((r) => r.data);

export const restoreAsset = (id: string) =>
  apiClient.post<Asset>(`/assets/${id}/restore`).then((r) => r.data);

/** Every location in use across the org's assets — the filter dropdown must offer all of
 *  them, not the handful that happen to appear on the page you are looking at. */
export const getAssetLocations = () =>
  apiClient.get<string[]>("/assets/locations").then((r) => r.data);

export const getAssetSummary = () =>
  apiClient.get<AssetSummary>("/assets/summary").then((r) => r.data);

export const createAsset = (data: AssetInput) =>
  apiClient.post<Asset>("/assets", data).then((r) => r.data);

export const updateAsset = (id: string, data: Partial<AssetInput>) =>
  apiClient.patch<Asset>(`/assets/${id}`, data).then((r) => r.data);

export const deleteAsset = (id: string) =>
  apiClient.delete(`/assets/${id}`).then((r) => r.data);

// Re-send the receipt-confirmation email to the current assignee.
export const resendAcknowledgement = (id: string) =>
  apiClient.post<Asset>(`/assets/${id}/resend-acknowledgement`).then((r) => r.data);
